RavenDB-11892 Memory leak in ByteStringContext

karmeli87 opened the pull request:
Resetting the context doesn't free the memory as long as we had only 2 segments.

So with many small allocation, this caused to the hot segments to grow to their maximal value of 16 MB.
And in presence of many contexts we ended up in a high memory usage.

Since ByteStringContext is mostly used for converting from string to slice document ids and different keys for voron tress/indexes, we do not expect to have large segments.

Therefor we changed the default size of minimum block from 64k to 4k and the maximal segment size to 2MB.

karmeli87 committed:
RavenDB-11892 Memory leak in ByteStringContext

Resetting the context doesn't free the memory as long as we had only 2 segments.

So with many small allocation, this caused to the hot segments to grow to their maximal value of 16 MB.
And in presence of many contexts we ended up in a high memory usage.

Since ByteStringContext is mostly used for converting from string to slice document ids and different keys for voron tress/indexes, we do not expect to have large segments.

Therefor we changed the default size of minimum block from 64k to 4k and the maximal segment size to 2MB.

karmeli87 commented:
@redknightlois can you please review this.

ayende commented on src/Sparrow/ByteString.cs:
```diff
@@ -626,7 +626,7 @@ public class SegmentStack : StackHeader<UnmanagedGlobalSegment>
     public sealed class ByteStringContext : ByteStringContext<ByteStringMemoryCache>
     {
-        public const int MinBlockSizeInBytes = 64 * 1024; // If this is changed, we need to change also LogMinBlockSize.
+        public const int MinBlockSizeInBytes = 4 * 1024; // If this is changed, we need to change also LogMinBlockSize.
```
@redknightlois Can you suggest what change is required in `LogMinBlockSize`?

redknightlois replied:
The calculation is Log(2, 4*1024) ... those are the actual potential buckets you could have given the minimum block size in bytes we are defining.

ayende replied:
Okay, so we should change it to 12

redknightlois commented on src/Sparrow/ByteString.cs:
```diff
@@ -809,7 +809,7 @@ private ByteString AllocateExternal(byte* valuePtr, int size, ByteStringType typ
                 {
-                    var tmp = Math.Min(16 * Constants.Size.Megabyte, _allocationBlockSize * 2);
+                    var tmp = Math.Min(2 * Constants.Size.Megabyte, _allocationBlockSize * 2);
                     AllocateExternalSegment(tmp);
```
No biggy here, we are reserving less (which is essentially what I am doing on the new allocators).

redknightlois commented on src/Sparrow/ByteString.cs:
```diff
@@ -735,9 +735,6 @@ public void Reset()
-            if (_wholeSegments.Count == 2)
-                return; // nothing to do
-
             Array.Clear(_internalReusableStringPoolCount, 0, _internalReusableStringPoolCount.Length);
```
This is the actual issue, we are not resetting the pointers. I wouldnt perform both changes simultaneously (fix + size change) because you can be impacting performance hard... Fix the issue where you reset the actual pointers, check if the problem continue to be repro and if that is the case then perform the size change.

myarichuk replied:
We cannot reset pointers since we have places where we want to keep pointers until a call to Reset().
For example [here](https://github.com/ravendb/ravendb/blob/v4.1/src/Voron/Data/Tables/TableSchema.cs#L366) - we want to keep the Slice alive until the transaction commits (and the context is reset when it is returned to the pool)

redknightlois replied:
But this is under `Reset`, therefore after that the memory is not guaranteed to exist.

ayende replied:
Yes, but the problem is that callers have allocated memory and did _not_ give it back.
We have to reset the state of the allocator back to the original

redknightlois commented on src/Sparrow/ByteString.cs:
```diff
@@ -1013,6 +1013,7 @@ private ByteString AllocateWholeSegment(int length, ByteStringType type)
             var byteString = Create(segment.Current, length, segment.Size, type);
             segment.Current += byteString._pointer->Size;
+            _currentlyAllocated += size;
 
```
`_currentlyAllocated` only track whole segments (which are the only actual memory allocated from the OS). The only place where `_currentlyAllocated` should count is there. To me this is not the case.

karmeli87 replied:
As I see it we are using this value only for debugging.. But even so `AllocateWholeSegment` calls `AllocateSegment(size)` so it is an actual allocation.

redknightlois replied:
What I mean is that the only place where `_currentlyAllocated` must be updated is right after we allocate on `AllocateSegment(size)` when it hits the actual allocator...

redknightlois reviewed (commented):
EDIT: Deleted

karmeli87 committed:
RavenDB-11892 Memory leak in ByteStringContext

Fix PR comments

karmeli87 committed:
RavenDB-11892 Memory access violation due to using a partial segment which his unmanaged segment was freed

When we reset the context we need to make sure that the internal segment is not a partial one,
but whole segment that is linked with an unmanaged segement.
